### DISCUSSION TITLE
Should "filename" parameter in `Content-Disposition: form-data` be mandatory???

### DISCUSSION THREAD

**Discussion by memsuser** [encode/starlette#1703] [Potential Issue] [answered]
As I see now in Starlette code, the parameter "filename" is mandatory to generate **UploadFile**. Otherwise data will be parsed as a string. Is this an intended behaviour? 

RFC-7578 says:
 > For form data that represents the content of a file, a name for the
   file **SHOULD** be supplied as well, by using a "filename" parameter of
   the Content-Disposition header field.  The file name isn't mandatory
   for cases where the file name isn't available or is meaningless or
   private; this might result, for example, when selection or drag-and-
   drop is used or when the form data content is streamed directly from
   a device.

So when I used just "name" field without "filename", the result was a bit unexpected, and forced me to look up sources. 
Probably this nuance just should be documented.
 
The code I'm talking about is listed below:
  
```python
# https://github.com/encode/starlette/blob/1225ae8404ed0ae0b9170fde1cf824ec3f9c0f2f/starlette/formparsers.py#L221
# ...
elif message_type == MultiPartMessage.HEADERS_FINISHED:
    disposition, options = parse_options_header(content_disposition)
    try:
        field_name = _user_safe_decode(options[b"name"], charset)
    except KeyError:
        raise MultiPartException(
            'The Content-Disposition header field "name" must be '
            "provided."
        )
    if b"filename" in options:
        filename = _user_safe_decode(options[b"filename"], charset)
        file = UploadFile(
            filename=filename,
            content_type=content_type.decode("latin-1"),
            headers=Headers(raw=item_headers),
        )
    else:
        file = None
    elif message_type == MultiPartMessage.PART_DATA:
    if file is None:
        data += message_bytes
    else:
        await file.write(message_bytes)
    elif message_type == MultiPartMessage.PART_END:
    if file is None:
        items.append((field_name, _user_safe_decode(data, charset)))
    else:
        await file.seek(0)
        items.append((field_name, file))
```

**Comment by adriangb** [comment]
The issue is that there is no other way (that I know of) to differentiate files from forms. But yes the docs could use some work in this area. Would you be open to submitting a PR? I'm happy to coach you through it.

**Comment by memsuser** [comment]
> The issue is that there is no other way (that I know of) to differentiate files from forms

Ah, I see. Then, I don’t know either. 

Okay, I'll make a PR then.

**Reply by memsuser** [reply]
Made a PR: https://github.com/encode/starlette/pull/1704

**Comment by memsuser** [comment, marked as answer]
Gave it another thought, though.

Maybe we can use `Content-Type` header of the form-data part to interpret form-data as file? 

For example, such a request can be formed and seems to have unambiguous hint that there is a file inside with MIME-type `'image/jpeg'`:

```http
POST /detection HTTP/1.1
Host: localhost:7103
Content-Length: 281
Content-Type: multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW

----WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="drawing";
Content-Type: image/jpeg

(data)
----WebKitFormBoundary7MA4YWxkTrZu0gW
```

**Reply by adriangb** [reply]
I think if it that header is there then yes it's probably a file. But I don't think every client sends that. At least requests in Python does not:

```python
from io import BytesIO
from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import Response
from starlette.routing import Route
from starlette.testclient import TestClient

async def endpoint(request: Request) -> Response:
    body = await request.body()
    print(body)
    print(b"Content-Type" in body)
    return Response(status_code=200)


app = Starlette(routes=[Route("/", endpoint, methods=["POST"])])

client = TestClient(app)

resp = client.post("/", data={"bar-form": b"foo bar"}, files={"foo.txt": BytesIO(b"foo")})
assert resp.status_code == 200, resp.content
```

**Reply by memsuser** [reply]
Oh, I see. Then it seems like a necessary decision. 
The documentation should be plenty as long as the design goal is clear.
